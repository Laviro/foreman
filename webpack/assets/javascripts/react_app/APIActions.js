import { API_OPERATIONS } from './APIConstants';

export const stopPollingAction = key => ({
  type: API_OPERATIONS.STOP_POLLING,
  payload: { key },
});
