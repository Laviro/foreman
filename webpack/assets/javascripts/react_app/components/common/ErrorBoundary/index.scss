.error-boundary-foreman-app {
  .empty-state-icon {
    color: var(--pf-global--palette--red-200);
  }
}
