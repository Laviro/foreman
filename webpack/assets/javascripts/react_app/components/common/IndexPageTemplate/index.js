import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import Table from '../../PF4/Table';
import PageLayout from '../../../routes/common/PageLayout/PageLayout';
import { fetchData } from '../../PF4/Table/TableActions';
import { getControllerSearchProps } from '../../../constants';
import './index.scss';
import { selectSearch, selectResponse } from '../../PF4/Table/TableSelectors';
import { getTableAPIKey } from '../../PF4/Table/TableHelpers';

const IndexPageTemplate = ({
  path,
  layoutProps,
  tableProps,
  paginationProps,
}) => {
  const dispatch = useDispatch();
  const searchQuery = useSelector(selectSearch);
  const { title, header } = useSelector(state =>
    selectResponse(state, getTableAPIKey(path))
  );
  const onSearch = nextQuery =>
    dispatch(fetchData(path, { query: nextQuery, page: 1 }));
  return (
    <PageLayout
      className="index_page_template"
      searchable
      searchQuery={searchQuery}
      searchProps={getControllerSearchProps(path)}
      onSearch={onSearch}
      header={header}
      title={title}
      {...layoutProps}
    >
      <Table
        requestPath={path}
        {...tableProps}
        paginationProps={paginationProps}
      />
    </PageLayout>
  );
};

export default IndexPageTemplate;
