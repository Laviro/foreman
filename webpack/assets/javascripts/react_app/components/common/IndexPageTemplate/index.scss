#main.index_page_template {
  padding: 0;

  h1 {
    font-weight: bold;
    margin-bottom: 0;
  }

  #index-page-top-pagination {
    margin-right: 24px;
  }

  .index-page-description {
    white-space: nowrap;
    margin-bottom: 15px;
  }
}
