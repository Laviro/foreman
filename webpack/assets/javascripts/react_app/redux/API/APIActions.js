import { API_OPERATIONS } from './APIConstants';

export const stopPolling = key => ({
  type: API_OPERATIONS.STOP_POLLING,
  key,
});

export const startPolling = (key, APIRequest, polling) => ({
  type: API_OPERATIONS.START_POLLING,
  key,
  payload: {
    APIRequest,
    polling,
  },
});

export default {
  startPolling,
  stopPolling,
};
