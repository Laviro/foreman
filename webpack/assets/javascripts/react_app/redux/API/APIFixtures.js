import Immutable from 'seamless-immutable';
import { noop } from '../../common/helpers';

export const key = 'SOME_KEY';
export const APIRequest = noop;
export const polling = 3000;
export const initialState = Immutable({
  polling: {},
});
export const stateWithKey = Immutable({
  polling: {
    [key]: 1,
  },
});
