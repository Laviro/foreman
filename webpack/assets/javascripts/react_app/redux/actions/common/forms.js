import { APIActions } from '../../API';

import { addToast } from '../toasts';
import { sprintf, translate as __ } from '../../../../react_app/common/I18n';

export class SubmissionError {
  constructor(errors) {
    this.errors = errors;
  }
}

const fieldErrors = ({ error }) => {
  const { errors, severity } = error;

  if (errors.base) {
    errors._error = {};
    errors._error.errorMsgs = errors.base;
    errors._error.severity = severity;
    delete errors.base;
  }

  return new SubmissionError(errors);
};

export const onError = error => {
  if (error.response.status === 422) {
    // Handle invalid form data
    throw fieldErrors(error.response.data);
  }
  throw new SubmissionError({
    _error: {
      errorMsgs: [
        `${__('Error submitting data:')} ${error.response.status} ${__(
          error.response.statusText
        )}`,
      ],
    },
  });
};

const verifyProps = (item, values) => {
  if (!item) {
    throw new Error('item must be defined, e.g. Bookmark');
  }
  if (!values) {
    throw new Error('values must be defined');
  }
};

export const submitForm = ({
  item,
  url,
  values: params,
  message,
  method = 'post',
  headers,
  apiActionTypes: actionTypes,
}) => {
  verifyProps(item, params);
  return async dispatch => {
    const uniqueAPIKey = `${item.toUpperCase()}_FORM_SUBMITTED`;

    const handleError = error => onError(error);

    const handleSuccess = ({ data }) => {
      dispatch({
        type: uniqueAPIKey,
        payload: { item, data },
      });
      dispatch(
        addToast({
          type: 'success',
          // eslint-disable-next-line no-undef
          message: message || sprintf('%s was successfully created.', __(item)),
        })
      );
    };

    dispatch(
      APIActions[method]({
        key: uniqueAPIKey,
        url,
        headers,
        params,
        actionTypes,
        handleError,
        handleSuccess,
      })
    );
  };
};
