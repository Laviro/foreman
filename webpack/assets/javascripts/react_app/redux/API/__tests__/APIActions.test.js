import { testActionSnapshotWithFixtures } from '../../../common/testHelpers';
import { startPolling, stopPolling } from '../APIActions';
import { key, APIRequest, polling } from '../APIFixtures';

const fixtures = {
  'should start polling': () => startPolling(key, APIRequest, polling),
  'should stop polling': () => stopPolling(key),
};

describe('API actions', () => testActionSnapshotWithFixtures(fixtures));
