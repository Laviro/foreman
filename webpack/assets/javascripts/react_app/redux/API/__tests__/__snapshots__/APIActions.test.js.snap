// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API actions should start polling 1`] = `
Object {
  "key": "SOME_KEY",
  "payload": Object {
    "APIRequest": [Function],
    "polling": 3000,
  },
  "type": "API_START_POLLING",
}
`;

exports[`API actions should stop polling 1`] = `
Object {
  "key": "SOME_KEY",
  "type": "API_STOP_POLLING",
}
`;
